tests/integration/test_healing.py

```diff
@@ -0,0 +1,281 @@
+
+    # freeze (STOP signal) mysqld for the unit
+    logger.info(f"Freezing process id {pid}")
+    await ops_test.juju("ssh", unit.name, "sudo", "kill", "-19", pid)
+
+    config = {
+        "username": CLUSTER_ADMIN_USERNAME,
+        "password": await get_system_user_password(unit, CLUSTER_ADMIN_USERNAME),
+        "host": await unit.get_public_address(),
+    }
+
+    # verify that connection is not possible
+    assert not is_connection_possible(config), "❌ Mysqld is not paused"
+
+    # unfreeze (CONT signal) mysqld for the unit
+    logger.info(f"Unfreezing process id {pid}")
+    await ops_test.juju("ssh", unit.name, "sudo", "kill", "-18", pid)
+
+    # verify that connection is possible
+    assert is_connection_possible(config), "❌ Mysqld is not paused"
+
+    # verify instance is part of the cluster
+    logger.info("Check if instance in cluster")
+    assert await is_unit_in_cluster(
+        ops_test, unit.name, another_unit.name
+    ), "❌ Unit not online in the cluster"
+
+
+@pytest.mark.order(3)
+@pytest.mark.abort_on_fail
+@pytest.mark.healing_tests
+async def test_network_cut(ops_test: OpsTest):
+    """Completely cut and restore network."""
+    app = await app_name(ops_test)
+    unit = ops_test.model.applications[app].units[0]
+    another_unit = ops_test.model.applications[app].units[1]
+
+    # get unit hostname
+    hostname = await unit_hostname(ops_test, unit.name)
+
+    logger.info(f"Unit {unit.name} it's on machine {hostname} ✅")
+
+    model_name = ops_test.model.info.name
+    unit_ip = instance_ip(model_name, hostname)
+
+    config = {
+        "username": CLUSTER_ADMIN_USERNAME,
+        "password": await get_system_user_password(unit, CLUSTER_ADMIN_USERNAME),
+        "host": unit_ip,
+    }
+
+    # verify that connection is possible
+    assert is_connection_possible(config), f"❌ Connection to host {unit_ip} is not possible"
+
+    logger.info(f"Cutting network for {hostname}")
+    cut_network_from_unit(hostname)
+
+    # verify that connection is not possible
+    assert not is_connection_possible(config), "❌ Connection is possible after network cut"
+
+    logger.info(f"Restoring network for {hostname}")
+    restore_network_for_unit(hostname)
+
+    # wait until network is reestablished for the unit
+    wait_network_restore(model_name, hostname, unit_ip)
+
+    # update instance ip as it may change on network restore
+    config["host"] = instance_ip(model_name, hostname)
+
+    # verify that connection is possible
+    assert is_connection_possible(config), "❌ Connection is not possible after network restore"
+
+    # verify instance is part of the cluster
+    logger.info("Check if instance in cluster")
+    assert await is_unit_in_cluster(
+        ops_test, unit.name, another_unit.name
+    ), "Unit not online in the cluster"
+
+
+@pytest.mark.order(4)
+@pytest.mark.abort_on_fail
+@pytest.mark.healing_tests
+async def test_replicate_data_on_restart(ops_test: OpsTest):
```

Comment:
might be good to combine this with the other test that verifies restart. Also you will want to remove any data you created here since it is possible someone passed in a provided cluster.

Reply:
yep. I'll drop the table to ensure it's always clean

Reply:
this is a thorough test, at some point you'll need to verify that all ops (network cut, kill, restart, etc) correctly replicate data